src/rgw/rgw_rest_s3.cc

```diff
@@ -6259,31 +6431,59 @@ int RGWSelectObj_ObjStore_S3::run_s3select(const char* query, const char* input,
+  { //error-flow (syntax-error)
+    m_aws_response_handler.get()->send_error_response(s3select_syntax_error,
+                                                      s3select_syntax->get_error_description().c_str(),
+                                                      s3select_resource_id);
 
-  if (s3select_syntax->get_error_description().empty() == false) {
-    m_result.append(s3select_syntax->get_error_description());
-    ldpp_dout(this, 10) << "s3-select query: failed to prase query; {" << s3select_syntax->get_error_description() << "}"<< dendl;
-    status = -1;
+    ldpp_dout(this, 10) << "s3-select query: failed to prase query; {" << s3select_syntax->get_error_description() << "}" << dendl;
+    return -1;
   }
-  else {
-    status = m_s3_csv_object->run_s3select_on_stream(m_result, input, input_length, s->obj_size);
-    if(status<0) {
-      m_result.append(m_s3_csv_object->get_error_description());
+  else
+  {
+
+    m_aws_response_handler.get()->init_success_response();
+    length_before_processing = (m_aws_response_handler->get_sql_result()).size();
+
+    //query is correct(syntax), processing is starting.
+    status = m_s3_csv_object->run_s3select_on_stream(m_aws_response_handler.get()->get_sql_result(), input, input_length, s->obj_size);
+    length_post_processing = (m_aws_response_handler->get_sql_result()).size();
+    m_aws_response_handler.get()->set_total_bytes_returned(length_post_processing-length_before_processing);
+    if (status < 0)
+    { //error flow(processing-time)
+      m_aws_response_handler.get()->send_error_response(s3select_processTime_error,
+                                                        m_s3_csv_object->get_error_description().c_str(),
+                                                        s3select_resource_id);
+
+      ldpp_dout(this, 10) << "s3-select query: failed to process query; {" << m_s3_csv_object->get_error_description() << "}" << dendl;
+      return -1;
     }
-  }
 
-  if (m_result.size() > strlen(PAYLOAD_LINE)) {
-    m_result.append(END_PAYLOAD_LINE);
-    int buff_len = create_message(m_result, m_result.size() - 12, header_size);
-    s->formatter->write_bin_data(m_result.data(), buff_len);
-    if (op_ret < 0) {
-      return op_ret;
+    if (chunk_number == 0)
+    {//success flow
+      if (op_ret < 0)
+      {
+        set_req_state_err(s, op_ret);
+      }
+      dump_errno(s);
+      // Explicitly use chunked transfer encoding so that we can stream the result
+      // to the user without having to wait for the full length of it.
+      end_header(s, this, "application/xml", CHUNKED_TRANSFER_ENCODING);
     }
+    chunk_number++;
+  }
+  
+  if (length_post_processing-length_before_processing != 0) {
+    m_aws_response_handler.get()->send_success_response();
+  } else {
+    m_aws_response_handler.get()->init_continuation_response();
   }
-  rgw_flush_formatter_and_reset(s, s->formatter);
+
+  m_aws_response_handler.get()->init_progress_response();
```

Comment:
please add a conditional flag as described in https://docs.aws.amazon.com/cli/latest/reference/s3api/select-object-content.html (--request-flag)